geps/gep-2162.md

```diff
@@ -0,0 +1,114 @@
+
+With the evolution of the Conformance tests suite and features, we want to;
+
+1. Enable users easily see what features the implementation support
+1. Improving UX and potentially provide clear errors when a user is trying to use a feature the installed GWC doesn't support
+
+This doc proposes to enhance the GatewayClassStatus API so implementations could publish a list of features they support/don't support.
+
+Implementations **should** publish the supported/unsupported features before Accepting the GatewayClass.
+
+They can either maintain static lists of supported/unsupported features OR generate if from their conformance tests results/reports as part of their CI.
+
+## API
+
+This GEP proposes API changes describes as follow:
+1. Adding `FeaturesSupport` and  `GatewayClassFeatures` structs. 
+2. Update the `GatewayClassStatus` struct to include `GatewayClassFeatures`
+
+```go
+// FeaturesSupport contains the supported features and unsupported features by the installed GatewayClass.
+type FeaturesSupport struct {
+	// SupportedFeatures indicates which extended features were flagged as
+	// supported by the implementation and tests will be attempted for.
+	SupportedFeatures []string `json:"supportedFeatures,omitempty"`
+
+	// UnsupportedFeatures indicates which extended features the implementation
+	// does not have support for and therefore will not attempt to test.
+	UnsupportedFeatures []string `json:"unsupportedFeatures,omitempty"`
+}
+
+// GatewayClassFeatures is the features supported by the GatewayClass broken down by support level.
+type GatewayClassFeatures struct {
+	// Core indicates the core support level features. UnsupportedFeatures should be null for the implementation to be considered all conformant. 
+    Core FeaturesSupport `json:"core,omitempty"`
+
+	// Extended indicates the extended support level features.
+    Extended FeaturesSupport `json:"extended,omitempty"`
+    
+}
+// GatewayClassStatus is the current status for the GatewayClass.
+type GatewayClassStatus struct {
+	// Conditions is the current status from the controller for
+	// this GatewayClass.
+	//
+	// Controllers should prefer to publish conditions using values
+	// of GatewayClassConditionType for the type of each Condition.
+	//
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	// +kubebuilder:validation:MaxItems=8
+	// +kubebuilder:default={{type: "Accepted", status: "Unknown", message: "Waiting for controller", reason: "Pending", lastTransitionTime: "1970-01-01T00:00:00Z"}}
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+    // Features is the features supported by the GatewayClass broken down by support level.
+    Features GatewayClassFeatures `json:"features,omitempty"`
+}
+```
+
+## Conformance Details
+
+(TBD)
```

Comment:
Is the intention to do this as a separate follow up PR?

Reply:
not sure about that yet. I think we will not be able to test this but will revisit once the GEP is ready to merge

Reply:
I don't think we need much here other than just a statement that conformance tests will read this information from GatewayClass unless it is specified in some other way.